Docs setup for the Task SDK

sunank200 opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->
This PR is a continuation of https://github.com/apache/airflow/pull/47357

To enhance the experience (i.e., rather than just being given an alphabetical list of classes/functions), I have opted not to use the AutoAPI extension's auto-generate feature, allowing me to precisely control the order and grouping of the classes.

I have organized Task-SDK API into these sections:
* **Defining DAGs**: `DAG`, `dag`
* **Decorators**: `task`, `task_group`, `setup`, `teardown`
* **Bases**: `BaseOperator`, `BaseSensorOperator`, `BaseNotifier`, `BaseOperatorLink`, `XComArg`
* **Connections & Variables**: `Connection`, `Variable`
* **Tasks & Operators**: `TaskGroup`, `get_current_context`, `get_parsing_context`, `Param`, `chain`, `chain_linear`, `cross_downstream`
* **Assets**: `Asset`, `AssetAlias`, `AssetAll`, `AssetAny`, `AssetWatcher`, `Metadata`

I also ensured that all documented references use the top-level SDK path (e.g., `airflow.sdk.DAG`) instead of exposing the underlying module paths (e.g., `airflow.sdk.definitions.dag.DAG`).

I have incorporated a test that compares the items in the generated `objects.inv` with the elements I re-export from `airflow/sdk/__init__.py`.

The `airflow/sdk/__init__.pyi` type stub is unfortunately necessary, as without it (and with the imports solely within the `if TYPE_CHECKING` block), the re-exported classes were not appearing where I wanted them. Specifically, I want users to import items directly from `airflow.sdk`, not the submodule.

        stubgen -o task-sdk/src -m airflow.sdk

Test this with

     breeze build-docs task-sdk

or 

    uv run --group docs build-docs task-sdk

**Screenshots**
<img width="1720" alt="Screenshot 2025-06-02 at 6 15 40 PM" src="https://github.com/user-attachments/assets/7ce06d9f-1af3-4156-acf7-642d5f37f907" />
<img width="1726" alt="Screenshot 2025-06-02 at 7 04 10 PM" src="https://github.com/user-attachments/assets/adfa99b3-1c40-4c37-b523-4d6ee27381b2" />
<img width="1707" alt="Screenshot 2025-06-02 at 6 44 01 PM" src="https://github.com/user-attachments/assets/d5ccbabc-dfbd-465d-ae32-3697acdce827" />
<img width="1728" alt="Screenshot 2025-06-11 at 1 06 26 AM" src="https://github.com/user-attachments/assets/c8a5792b-59ad-4855-8937-4877a31a9a55" />




<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->
closes: [#43010](https://github.com/apache/airflow/issues/43010)
closes: [#51283](https://github.com/apache/airflow/issues/51283)

<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

ashb committed:
Start working on the docs for the Task SDK

In order to have a nicer experience (i.e. rather than just being given an
alphabetical list of classes/functions) I have chosen to not use the auto
generate feature of the AutoAPI extension, but instead to precisely control
the order and grouping of the classes.

For this to be complete we will likely need some testing that compares the
items on the generated `objects.inv` with the things we re-export from
`airflow/sdk/__init__.py`.

The `airflow/sdk/__init__.pyi` type stub is sadly needed, as without it (and
with the imports only inside the `if TYPE_CHECKING` block) we weren't getting
the re-exported classes showing up where we want them, specifically we want
users to import things directly from airflow.sdk, not the submodule.

It was generated with `stubgen -o task_sdk/src -m airflow.sdk`

Test this with `sphinx-build -b html -Tv task_sdk/docs/ task_sdk/docs/_build`

sunank200 committed:
Add Breeze build-task-sdk-docs command and use the top-level SDK path (e.g. airflow.sdk.DAG)

fix static checks

Fix the static checks

Fix the tests

Fix autoapimodule

explicitly import all your public API into airflow/sdk/__init__.py (including literal, MappedOperator, SecretCache, etc.), and we wired up __lazy_imports and __all__ so AutoAPI only sees those names.

Add SecretCache in the SDK module for both runtime and static typing

use the top-level SDK path (e.g. airflow.sdk.DAG)

Add Breeze build-task-sdk-docs command; tests for Task-SDK API vs docs; clean up Sphinx config

kaxil commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,77 @@
+   :members:
+
+
+Tasks and Operators
```
We should also have  `task` decorator, `Param`, `setup`, `teardown` etc

sunank200 replied:
I have added a section for this.

ashb commented on task-sdk/docs/conf.py:
I'm back off vacation tomorrow, I'll take a look in my morning 

Please don't merge before then

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,161 @@
+   :language: python
+   :start-after: [START simplest_dag]
+   :end-before: [END simplest_dag]
+   :caption: Simplest DAG with ``@dag`` and ``@task``
```
Can we make these links?
```suggestion
   :caption: Simplest DAG with :func:`@dag <airflow.sdk.dag>` and :func:`@task <airflow.sdk.task>`
```

sunank200 replied:
Changed it.

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,161 @@
+   :language: python
+   :start-after: [START dag_decorator_usage]
+   :end-before: [END dag_decorator_usage]
+   :caption: Using the ``@dag`` decorator with custom tasks and operators.
```
Ditto here -- link please

sunank200 replied:
changed it

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,161 @@
+------------
+Defining DAGs
+~~~~~~~~~~~~~
+Use ``@dag`` to convert a function into an Airflow DAG. All nested ``@task`` calls
```
These ones probably don't need to be links, so should stay as they are.

sunank200 replied:
Why is the link not needed here?

ashb replied:
Because we've already linked to these earlier up on the page? I don't mind if you think these are better as links

sunank200 replied:
I have removed the links as suggested.

ashb reviewed (commented):
We are missing a dep on `apache-airflow-devel-common[docs]` to pull in the sphinx deps correctly

I'm not sure if it's just me locally, but there is a huge amount of whitespace all over the shop.

<img width="2130" alt="Screenshot 2025-06-03 at 10 05 05" src="https://github.com/user-attachments/assets/1792f169-4518-4e65-9752-5d68119b509c" />

I think that including all the example dags on the index page is too much, and we should move some/most of these to another page instead.

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,159 @@
+
+.. autoapimodule:: airflow.sdk.execution_time.task_runner
+   :members:
+   :undoc-members:
```
These explicitly must not be documented -- they are not part of the public interface.

sunank200 replied:
Removed it

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,161 @@
+
+  * **Decorators**:
+    - ``@task``, ``@task_group``, ``@setup``, ``@teardown`` (exported at top level).
+    - These live in ``airflow.sdk.decorator`` but are re-exported in ``airflow.sdk``.
```
Don't even _tell_ users about `airflow.sdk.decorator` module.

sunank200 replied:
Removed it

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,161 @@
+  * **Execution Time** (under ``airflow.sdk.execution_time``):
+    - ``supervisor``
+    - ``task_runner``
+
```
```suggestion
```

sunank200 replied:
removed it.

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,161 @@
+
+  * **I/O** helpers for managing runtime I/O serialization.
+
+  * **API client** (under ``airflow.sdk.api.client``) for interacting with Airflow's REST endpoints.
```
```suggestion
```

sunank200 replied:
removed it.

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,161 @@
+
+Refer to `api.html`_ for the complete reference of all decorators and classes.
+
+.. _api.html: api.html
```
```suggestion
Refer to :doc:`api` for the complete reference of all decorators and classes.
```

sunank200 replied:
changed it

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,161 @@
+    - ``supervisor``
+    - ``task_runner``
+
+  * **I/O** helpers for managing runtime I/O serialization.
```
Not sure these are useful/worth documenting right now.

sunank200 replied:
removed it for now

ashb reviewed (commented):
More generally - we should link to the API ref  more often from the index page

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,159 @@
+.. autoapifunction:: airflow.sdk.teardown
+
+.. autofunction:: airflow.sdk.task
+.. autofunction:: airflow.sdk.setup
+.. autofunction:: airflow.sdk.teardown
```
Future work: these are missing doc strings or any indication of the params they take.

sunank200 replied:
I can create subsequent PRs for this.

sunank200 replied:
https://github.com/apache/airflow/issues/51518

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,159 @@
+
+.. autoapifunction:: airflow.sdk.teardown
+
+.. autofunction:: airflow.sdk.task
```
Future work: we need to document the "sub" decorators such as `@task.skip_if` etc.

sunank200 replied:
I can create subsequent PRs for this.

sunank200 replied:
https://github.com/apache/airflow/issues/51518

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,159 @@
+.. autoapifunction:: airflow.sdk.literal
+
+.. autoapifunction:: airflow.sdk.__version__
+.. autofunction:: airflow.sdk.__version__
```
a) Why do we have this twice?
b) it looks wrong:
<img width="1261" alt="Screenshot 2025-06-03 at 10 27 13" src="https://github.com/user-attachments/assets/0ef7163a-5b18-4607-b270-cea43ab3da88" />

c) do we even want to document this? I think not.

sunank200 replied:
a. `autoapifunction` and `autofunction` aren’t actually the same Sphinx directive. In  conf.py, I am loading both
- autoapi = static signature + link into the separate API‐tree
- autodoc = full, live doc‐string detail in place.

b. It looks wrong. I have removed it

sunank200 replied:
c. I was thinking of showing the version

ashb replied:
The rendered docs  for publishing will show the version selector

<img width="1122" alt="Screenshot 2025-06-03 at 14 24 54" src="https://github.com/user-attachments/assets/e9b3d5c6-e902-4b63-b2d7-3ace7097917a" />

so we don't need to try and show the version here.

sunank200 replied:
Makes sense.

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,159 @@
+
+.. autoapifunction:: airflow.sdk.cross_downstream
+
+.. autoapifunction:: airflow.sdk.literal
```
I'd put these "tasks and operators", possibly under a sub section of "setting dependencies" or something

sunank200 replied:
Moved it

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,159 @@
+----------
+.. autoapimodule:: airflow.sdk.api.client
+   :members:
+   :undoc-members:
```
```suggestion
```

sunank200 replied:
removed it

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,159 @@
+.. rubric:: Cache
+
+.. autoapiclass:: airflow.sdk.SecretCache
+.. autoclass:: airflow.sdk.SecretCache
```
Internal impl detail, don't document

sunank200 replied:
mypy-airflow-core was failing. That's the reason I added it: https://github.com/apache/airflow/actions/runs/15343037924/job/43173553034?pr=51153

sunank200 replied:
I've exposed MappedOperator in the SDK exactly like SecretCache so that both runtime imports and mypy picks it up

ashb replied:
Mypy is an entirely different issue to what appears in the rendered docs.

ashb replied:
And this error:

```

airflow-core/src/airflow/models/variable.py:34: error: Module "airflow.sdk" has
no attribute "SecretCache"  [attr-defined]
    from airflow.sdk import SecretCache
    ^
```

also means we should not import it from `airflow.sdk` directly, but from `airflow.sdk.execution_time...` etc.

sunank200 replied:
Removed it

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,159 @@
+
+Metadata
+--------
+.. autoapiclass:: airflow.sdk.Metadata
```
This should be part of assets I think, given it's doc string is `Metadata to attach to an AssetEvent.`

sunank200 replied:
Moved it to assets

ashb commented on task-sdk/src/airflow/sdk/__init__.py:
```diff
@@ -126,3 +132,8 @@ def __getattr__(name: str):
+
+def __dir__() -> list[str]:
+    """Override dir() to expose only the public API names and internal attributes."""
+    return sorted(__all__ + ["__getattr__", "__lazy_imports"])
```
Why is this function needed? I thought with `__all__` defined dir() would return that list

sunank200 replied:
`from airflow.sdk import *` only determines what is imported and does not affect what `dir()` displays. By default, `dir(module)` lists items in `module.__dict__`, not what is in `all`. With lazy loading, lazy names are not in `__dict__` until imported, which disrupts tooling that relies on `dir()` to discover APIs.

To fix this, I have overridden `dir` to include `sorted(all + ["getattr", "__lazy_imports"])`, so users see the complete public API immediately.

ashb replied:
But what is calling `dir()` on the module? IDEs don't use this -- they'll use the pyi stub.

Nothing outside of this module should need to know about these two things you added?

sunank200 replied:
Yes - outside of this file, nobody should ever import or rely on getattr or lazy_imports; they’re purely internal plumbing. All still drives what `from airflow.sdk import *` pulls in, and those two names will never be part of the public API.

I have removed `__dir__()` now.

ashb commented on task-sdk/src/airflow/sdk/bases/operator.py:
```diff
@@ -77,9 +77,7 @@
-    from airflow.sdk.definitions.context import Context
-    from airflow.sdk.definitions.dag import DAG
-    from airflow.sdk.definitions.taskgroup import TaskGroup
+    from airflow.sdk import DAG, Context, TaskGroup
```
Generally from wihthin the SDK we import it via the full path, not the helper.

What wasn't right with this as they were?

sunank200 replied:
It was part of your PR too: https://github.com/apache/airflow/pull/47357/files#diff-a521dd73fbcd6da9aac5e46d98cf2e640da5467cc1b66ea7c6a87c126be35a65

That's the reason I kept it like that.

ashb replied:
There's a reason my PR was draft still 😆 

Does it "break" without this?

sunank200 replied:
I have changed it back. Doesn't break it.

ashb commented on task-sdk/src/airflow/sdk/definitions/context.py:
```diff
@@ -118,6 +118,8 @@ class AirflowParsingContext(NamedTuple):
     If these values are not None, they will contain the specific DAG and Task ID that Airflow is requesting to
     execute. You can use these for optimizing dynamically generated DAG files.
+
+    You can obtain the current values via :py:func:`.get_parsing_context`.
```
Note: this class doesn't appear in any of the docs. It probably should since we list this as a return type.

sunank200 replied:
<img width="553" alt="Screenshot 2025-06-09 at 5 20 04 PM" src="https://github.com/user-attachments/assets/381c897f-cfeb-4e29-b5b9-eed4b2433720" />
It is showing here.

ashb replied:
Yes -- my point was that that is plaIn text, not a link to the class about what is in it. Given it's just dag_id and task_id we could get away with just documenting the properties here in prose

sunank200 committed:
Fix PR comments.

Add imports in example DAGs, fix spell checks in task-sdk, trigger doc build only on doc file changes; compare by file path, line number, context, spelling, and message; ensure context_line is a string, add spellcheck, nuke mapped operator from api.rst, remove skip_util_classes and setup from conf, fix PR comments, revert changes for SecretCache and MappedOperator, remove SecretCache and MappedOperator, fix tests and remove dir() from __init__, add breeze build-docs task-sdk, revert build-task-sdk-docs docs and related images.

potiuk commented on dev/breeze/src/airflow_breeze/commands/developer_commands.py:
```diff
@@ -1153,6 +1153,35 @@ def autogenerate(
     fix_ownership_using_docker()
 
 
+@main.command(name="build-task-sdk-docs", help="Build the Apache Airflow Task Execution SDK documentation.")
```
Question: is there a particular reason why we want to have a separate command here rather than using the same mechanisms we already use for other distributions?

After the refactoring and extraction two months ago, it should be rather easy to simply use the same mechanisms as for example "airflow-core" users - where all the common code is in "devel-common" and things liike `uv run --group docs build-docs --autobuild`

This includes generation of the the docs in `generated` folder where they are later published using "publish-docs" command to S3. I have a feeling  that simply adding `task-sdk` a another documentation module would be simply easier when it comes to all the "auxiliary" steps of releasing and publishing. 

I am happy to give the right pointers if needed.

sunank200 replied:
I think I haven't wired the task-sdk distribution into the generic docs workflow that Breeze already provides for core/providers, so the quickest way to get task-sdk docs building was to add in a little standalone build-task-sdk-docs command.

sunank200 replied:
I could do it in task-sdk exactly the same flow by:
1. Extending the "docs" group in `devel-common` (or in Breeze's `developer_commands.py`) so that it knows about a `task-sdk` module and points it at the right source directory.
2. Allowing you to say `uv run --group docs build-docs --distribution task-sdk [--autobuild]` instead of running a `breeze build-task-sdk-docs`.

 That would let us remove the custom command entirely and keep everything under the one uv run --group docs.

potiuk replied:
Yes. That +:

> 1. Extending the "docs" group in `devel-common` (or in Breeze's `developer_commands.py`) so that it knows about a `task-sdk` module and points it at the right source directory.

I do not think you need to change the group in devel-common (unless you need some additional dependencies) - I think it is enough to add (in task-sdk pyproject.toml).

```
docs = [
    "apache-airflow-devel-common[docs]"
]
```

Then:

in global_constants.py:

```python
# packages that  providers docs
REGULAR_DOC_PACKAGES = [
    "apache-airflow",
    "docker-stack",
    "helm-chart",
    "apache-airflow-providers",
]
```

In docs_builder.py (in develo_common):

```python
    @property
    def _src_dir(self) -> Path:
        if self.package_name == "helm-chart":
            return AIRFLOW_CONTENT_ROOT_PATH / "chart" / "docs"
        if self.package_name == "apache-airflow":
            return AIRFLOW_CONTENT_ROOT_PATH / "airflow-core" / "docs"
        if self.package_name == "docker-stack":
            return AIRFLOW_CONTENT_ROOT_PATH / "docker-stack-docs"
        if self.package_name == "apache-airflow-providers":
            return AIRFLOW_CONTENT_ROOT_PATH / "providers-summary-docs"
        if self.package_name.startswith("apache-airflow-providers-"):
            package_paths = self.package_name[len("apache-airflow-providers-") :].split("-")
            return (AIRFLOW_CONTENT_ROOT_PATH / "providers").joinpath(*package_paths) / "docs"
        if self.package_name == "apache-airflow-ctl":
            return AIRFLOW_CONTENT_ROOT_PATH / "airflow-ctl" / "docs"
        console.print(f"[red]Unknown package name: {self.package_name}")
        sys.exit(1)
```
In docs_build.py

```python
def get_available_packages(include_suspended: bool = False, short_form: bool = False) -> list[str]:
    """Get list of all available packages to build."""
    provider_names = get_available_providers_distributions(include_suspended=include_suspended)
    if short_form:
        provider_names = [get_short_form(name) for name in provider_names]
    return [
        "apache-airflow",
        *provider_names,
        "apache-airflow-providers",
        "apache-airflow-ctl",
        "helm-chart",
        "docker-stack",
    ]
```

in selective_checks:

```python
        FileGroupForCi.DOC_FILES: [
            r"^docs",
            r"^devel-common/src/docs",
            r"^\.github/SECURITY\.md",
            r"^airflow-core/src/.*\.py$",
            r"^airflow-core/docs/",
            r"^providers/.*/src/",
            r"^providers/.*/tests/",
            r"^providers/.*/docs/",
            r"^providers-summary-docs",
            r"^docker-stack-docs",
            r"^chart",
            r"^task-sdk/src/",
            r"^airflow-ctl/src/",
            r"^airflow-core/tests/system",
            r"^airflow-ctl/src",
            r"^airflow-ctl/docs",
            r"^CHANGELOG\.txt",
            r"^airflow-core/src/airflow/config_templates/config\.yml",
            r"^chart/RELEASE_NOTES\.rst",
            r"^chart/values\.schema\.json",
            r"^chart/values\.json",
            r"^RELEASE_NOTES\.rst",
        ],
```

```python
    @cached_property
    def docs_list_as_string(self) -> str | None:
        _ALL_DOCS_LIST = ""
        if not self.docs_build:
            return None
        if self._default_branch != "main":
            return "apache-airflow docker-stack"
        if self.full_tests_needed:
            return _ALL_DOCS_LIST
        providers_affected = self._find_all_providers_affected(
            include_docs=True,
        )
        if (
            isinstance(providers_affected, AllProvidersSentinel)
            or "docs/conf.py" in self._files
            or "docs/build_docs.py" in self._files
            or self._are_all_providers_affected()
        ):
            return _ALL_DOCS_LIST
        packages = []
        if any(file.startswith(("airflow-core/src/airflow/", "airflow-core/docs/")) for file in self._files):
            packages.append("apache-airflow")
        if any(file.startswith("providers-summary-docs/") for file in self._files):
            packages.append("apache-airflow-providers")
        if any(file.startswith("chart/") for file in self._files):
            packages.append("helm-chart")
        if any(file.startswith("docker-stack-docs") for file in self._files):
            packages.append("docker-stack")
        if providers_affected:
            for provider in providers_affected:
                packages.append(provider.replace("-", "."))
        return " ".join(packages)
```

Likely some tests should be updated too.

Should be I think  all that is needed (maybe there is one or more place) - so that task-sdk docs are  build automatically and use the infrastructure - including publishing to s3 during releases - we just will need to update `dev/README-*.md` for release manager to also include `task-sdk` package when they are building docs.

potiuk replied:
It's just so much easier to run such doc builds with `uv run` now .... once I moved all the docs to docs subfolder and separated conf.py to be in docs folders separately for each distribution.

It's almost **pleasant** now to work with the docs with `uv run --autobuild` :D

potiuk replied:
And following the other question - up to you how you want to do it, but my suggestion is once the docs are build, try to add the changes above - and I am happy to help to sollve any issues

sunank200 replied:
I have added 

```
breeze build-docs task-sdk
```

and `uv run --group docs build-docs task-sdk` also works now. Please review @potiuk

potiuk commented on dev/breeze/src/airflow_breeze/commands/developer_commands_config.py:
```diff
@@ -22,6 +22,7 @@
         "start-airflow",
         "static-checks",
         "build-docs",
+        "build-task-sdk-docs",
```
```suggestion
```

sunank200 replied:
Removed it.

potiuk commented on task-sdk/docs/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+sphinx>=5.0
+sphinx-autoapi>=1.8
+sphinx-airflow-theme>=0.2.2
```
```suggestion
sphinx-airflow-theme@https://github.com/apache/airflow-site/releases/download/0.2.3/sphinx_airflow_theme-0.2.3-py3-none-any.whl
```

potiuk replied:
We use direct link now rather than puiblishing it in PyPI

sunank200 replied:
removed requirements.txt as we have pyproject

ashb commented on devel-common/src/docs/build_docs.py:
```diff
@@ -197,6 +197,16 @@ def perform_spell_check_for_single_package(build_specification: BuildSpecificati
     builder.is_autobuild = build_specification.is_autobuild
+    if build_specification.package_name == "task-sdk":
+        console.print(
+            f"[bright_blue]{build_specification.package_name:60}:[/] Skipping spelling check (not supported)"
```
Why not?

sunank200 replied:
It is complaining for `enchant` without it and when I installed https://pypi.org/project/pyenchant/ it was still complaining the same. So I had skipped it for now
```
task-sdk                                                         The above exception was the direct cause of the following exception:
task-sdk                                                         
task-sdk                                                         Traceback (most recent call last):
task-sdk                                                           File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/sphinx/cmd/build.py", line 414, in build_main
task-sdk                                                             app = Sphinx(
task-sdk                                                                   ^^^^^^^
task-sdk                                                           File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/sphinx/application.py", line 294, in __init__
task-sdk                                                             self.preload_builder(buildername)
task-sdk                                                           File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/sphinx/application.py", line 403, in preload_builder
task-sdk                                                             self.registry.preload_builder(self, name)
task-sdk                                                           File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/sphinx/registry.py", line 193, in preload_builder
task-sdk                                                             self.load_extension(app, entry_point.module)
task-sdk                                                           File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/sphinx/registry.py", line 544, in load_extension
task-sdk                                                             raise ExtensionError(
task-sdk                                                         sphinx.errors.ExtensionError: Could not import extension sphinxcontrib.spelling (exception: The 'enchant' C library was not found and maybe needs to be installed.
task-sdk                                                         See  https://pyenchant.github.io/pyenchant/install.html
task-sdk                                                         for details
task-sdk                                                         )
```

kaxil replied:
Add the following env var:

```shell
export PYENCHANT_LIBRARY_PATH=$(brew --prefix enchant)/lib/libenchant-2.dylib
```

if you install using brew:

```shell
brew update
brew install enchant
```

sunank200 replied:
Yes works now. Thanks!

sunank200 replied:
Added the spell-check

ashb commented on providers/openlineage/src/airflow/providers/openlineage/utils/utils.py:
```diff
@@ -68,9 +68,8 @@
-    from airflow.models import TaskInstance
+    from airflow.models import DAG, MappedOperator, TaskInstance
     from airflow.providers.common.compat.assets import Asset
-    from airflow.sdk import DAG, MappedOperator
```
This change looks unrelated. What's going on here?

sunank200 replied:
No, mypy is complaining about this in providers. That's the reason for the change.

sunank200 replied:
Here is the error: https://github.com/apache/airflow/actions/runs/15486821971/job/43603804201

sunank200 replied:
This is wrong, I realise. But even in lines 79- 85, it is doing the same. I will fix that too.

potiuk replied:
I think you should copy the try/except from the else clause. Looks like Openlineage provider is prepared to read MappedOperators from task.sdk when it's there and falls back to read both DAG an MappedOperator from the models - it does not do it for TYPE_CHECKING though.

````
if TYPE_CHECKING:
    from openlineage.client.event_v2 import Dataset as OpenLineageDataset
    from openlineage.client.facet_v2 import RunFacet, processing_engine_run
    from airflow.models import TaskInstance
    from airflow.sdk import DAG, MappedOperator
    from airflow.sdk.execution_time.secrets_masker import (
        Redactable,
        Redacted,
        SecretsMasker,
        should_hide_value_for_key,
    )
    from airflow.utils.state import DagRunState, TaskInstanceState
else:
    try:
        from airflow.sdk import DAG, MappedOperator
    except ImportError:
        from airflow.models import DAG, MappedOperator
```

And yeah - usually in such cases we separate such changes to a new PR (and it is fixed after rebasing the original one).

potiuk replied:
BTW. This change likely comes from the `task sdk's.pyi` that is added in this PR - so it proabably gives MyPy stronger signal of what is and what is not available as attribute during TYPE_CHECKING.

sunank200 replied:
Created a PR: https://github.com/apache/airflow/pull/51492

ashb commented on task-sdk/docs/_static/custom.css:
I don't think we should really need this -- we don't have it for any of the providers or airflow main do we?

sunank200 replied:
I was trying to see why there are a lot of spaces. Removing it.

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,134 @@
+-- we won't go out of our way break usage of them, but we make no promises either.
+
+.. :py:module: airflow.sdk
+
```
This is a comment right?
```suggestion
```

sunank200 replied:
Yes.. removed it.

ashb commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,134 @@
+.. autoapifunction:: airflow.sdk.literal
+
+Edges & Labels
+--------------
```
```suggestion
~~~~~~~~~~~~~~
```

so that it's a sub heading under Task & Ops? (I think that makes sense from a layout pov? Not 100% sure)

sunank200 replied:
This makes sense.

ashb commented on task-sdk/docs/conf.py:
```diff
@@ -0,0 +1,95 @@
+
+def setup(app):
+    # Skip utility classes from definitions modules, but expose DAG under airflow.sdk.DAG
+    app.connect("autoapi-skip-member", skip_util_classes)
```
Is this needed still?

sunank200 replied:
It can be removed because it is already re-exported, for example, DAG in __init__.py and listing it in __all__

sunank200 replied:
removed it

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,123 @@
+
+:any:`DAG` is where to start. :any:`dag`
+
+The Apache Airflow Task Execution SDK(Task SDK) provides Python-native interfaces for defining DAGs (via decorators),
```
Not just via decorators -- it has the DAG class too.

sunank200 replied:
Right. I have removed it.

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,123 @@
+
+Decorators
+~~~~~~~~~~
+Simplify DAG and task definitions using decorators:
```
We don't show the dag decorator here. Either add it, or remove DAG from the prose.

sunank200 replied:
I have removed DAG

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,123 @@
+Model data as assets and emit them to downstream tasks with the SDK's asset library under
+``airflow.sdk.definitions.asset``. You can use:
+
+- :func:`@asset <airflow.sdk.asset>`, :class:`AssetAlias <airflow.sdk.AssetAlias>`, etc. (see the **api reference** section below)
```
```suggestion
- :func:`@asset <airflow.sdk.asset>`, :class:`~airflow.sdk.AssetAlias`, etc. (see the **api reference** section below)
```

I think -- the only reason we need the "long form" for decorator is so that the link is `@asset` rather than just `asset()`, but for classes and fns the `~` prefix will make it use the last component automatically.

sunank200 replied:
Changed it.

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,123 @@
+    - ``XComArg``
+
+  * **Decorators**:
+    - :func:`@task <airflow.sdk.task>`, :func:`@task_group <airflow.sdk.task_group>`, :func:`@setup <airflow.sdk.setup>`, :func:`@teardown <airflow.sdk.teardown>`
```
Duplicates what's above

sunank200 replied:
Removed it

ashb commented on task-sdk/pyproject.toml:
```diff
@@ -213,3 +213,6 @@ pythonpath = "tests"
 tmp_path_retention_policy = "failed"
+
+[project.optional-dependencies]
+docs = ["apache-airflow-devel-common[docs]"]
```
Nit: this should be around L 146 so it's with the other deps

Additionally this should be a `dependency-groups`, as it doesn't ever make sense for a user to do `pip install "apache-airflow-task-sdk[docs]"` (which is what optional deps allows)

sunank200 replied:
changed it.

ashb commented on task-sdk/tests/test_docs_inventory.py:
```diff
@@ -0,0 +1,82 @@
+    return inv
+
+
+@pytest.mark.skipif(
+    shutil.which("sphinx-build") is None, reason="sphinx-build not available, skipping docs inventory test"
+)
```
This will fail if docs aren't built too right -- should it also test if the inventory file exists?

Do we somehow run this specific test as part of the docs build pipeline? Surely it can't pass via the normal CI test pipeline?

sunank200 replied:
Yes, the test will fail if the docs aren't built, because it explicitly tries to build the docs as part of the test:
```
subprocess.run([sphinx, "-b", "html", "-q", str(docs_dir), str(build_dir)], check=True)
inv_path = build_dir / "objects.inv"
assert inv_path.exists(), "objects.inv not found after docs build"
```
So, the test already checks for the inventory file’s existence after attempting to build the docs.

sunank200 replied:
<img width="1640" alt="Screenshot 2025-06-09 at 7 14 07 PM" src="https://github.com/user-attachments/assets/5be16a06-667b-499f-bc50-22094526d2ee" />
It is running as part of the CI

potiuk commented on dev/breeze/src/airflow_breeze/utils/selective_checks.py:
```diff
@@ -1194,6 +1194,8 @@ def docs_list_as_string(self) -> str | None:
             packages.append("helm-chart")
         if any(file.startswith("docker-stack-docs") for file in self._files):
             packages.append("docker-stack")
+        if any(file.startswith("task-sdk/src/") for file in self._files):
```
NIT: Also "^task-sdk/docs" should be added to 

```
        FileGroupForCi.DOC_FILES: [
```

Ony "^task-sdk/src" is there.

This way doc build will be triggered if only `doc` files are changed.

sunank200 replied:
Changed it

potiuk replied:
It's the ``        FileGroupForCi.DOC_FILES: [`` above that should be changed. Here it was fine.

sunank200 replied:
Makes sense. Changed it now.

kaxil commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,132 @@
+-----------------
+.. autoapiclass:: airflow.sdk.TaskGroup
+
+.. autoapiclass:: airflow.sdk.MappedOperator
```
Should nuke MappedOperator

sunank200 replied:
Yes. pushed just now

amoghrajesh commented on airflow-core/docs/index.rst:
```diff
@@ -32,6 +32,15 @@ Airflow workflows are defined entirely in Python. This "workflows as code" appro
+Task SDK
+========
+
+For Airflow Task SDK (Airflow 3.x+), see the standalone reference & tutorial site:
```
```suggestion
For Airflow Task SDK, see the standalone reference & tutorial site:
```

No need for this as the docs are versioned

sunank200 replied:
Changed it.

amoghrajesh commented on airflow-core/src/airflow/example_dags/example_assets.py:
```diff
@@ -50,6 +50,7 @@
 automatically as they depend on assets that do not get updated or are not produced by any scheduled tasks.
 """
 
+# [START asset_def]
```
Its ok to move the marker to after "from future" imports

sunank200 replied:
Changed it.

amoghrajesh commented on airflow-core/src/airflow/example_dags/example_asset_alias.py:
```diff
@@ -34,6 +34,7 @@
 
 import pendulum
 
+# [START example_asset_alias]
```
Marker doesn't cover the pendulum import

sunank200 replied:
Changed it.

amoghrajesh commented on airflow-core/src/airflow/example_dags/example_dag_decorator.py:
```diff
@@ -15,6 +15,7 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+# [START dag_decorator_usage]
```
nit: Move it after the from future import

sunank200 replied:
changed it.

amoghrajesh commented on dev/breeze/doc/03_developer_tasks.rst:
```diff
@@ -224,6 +224,11 @@ short ``provider id`` (might be multiple of them).
+.. code-block:: bash
+
+     # To build documentation for Task SDK package
+     breeze build-docs task-sdk
```
```suggestion
To build documentation for Task SDK package, use the below command
.. code-block:: bash

     breeze build-docs task-sdk
```

sunank200 replied:
changed it.

amoghrajesh commented on devel-common/src/sphinx_exts/docs_build/spelling_checks.py:
```diff
@@ -71,17 +71,17 @@ def __lt__(self, other):
         context_line_a: str = self.context_line or ""
         context_line_b: str = other.context_line or ""
-        left: tuple[Path, int, int, str, str] = (
+        left: tuple[Path, int, str, str, str] = (
```
Why this change?

sunank200 replied:
It actually is throwing the following error and its a bug:
```
==============================  task-sdk  ==============================
Traceback (most recent call last):
  File "/Users/sunank200/Documents/airflow/.venv/bin/build-docs", line 10, in <module>
    sys.exit(build_docs())
             ^^^^^^^^^^^^
  File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/click/core.py", line 1161, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/rich_click/rich_command.py", line 166, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/click/core.py", line 1443, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sunank200/Documents/airflow/.venv/lib/python3.11/site-packages/click/core.py", line 788, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/sunank200/Documents/airflow/devel-common/src/docs/build_docs.py", line 728, in build_docs
    print_build_errors_and_exit(
  File "/Users/sunank200/Documents/airflow/devel-common/src/docs/build_docs.py", line 428, in print_build_errors_and_exit
    display_spelling_error_summary(spelling_errors)
  File "/Users/sunank200/Documents/airflow/devel-common/src/sphinx_exts/docs_build/spelling_checks.py", line 157, in display_spelling_error_summary
    for warning_no, error in enumerate(sorted(errors), 1):
                                       ^^^^^^^^^^^^^^
  File "/Users/sunank200/Documents/airflow/devel-common/src/sphinx_exts/docs_build/spelling_checks.py", line 88, in __lt__
    return left < right
           ^^^^^^^^^^^^
TypeError: '<' not supported between instances of 'str' and 'int'
```

sunank200 replied:
`airflow‐core` docs build never actually hits the buggy sort, because there are no real spelling errors (or only a single one, which never forces a comparison). `task-sdk` build produces multiple spelling errors, one of which has no context line, so the sort tries to compare a string to an integer and blows up.

amoghrajesh replied:
Gotcha, thanks

potiuk commented on task-sdk/docs/examples.rst:
```diff
@@ -0,0 +1,86 @@
+
+Define a basic DAG and task in just a few lines of Python:
+
+.. literalinclude:: ../../airflow-core/src/airflow/example_dags/example_simplest_dag.py
```
NIT: I suggest to use `exampleinclude` for those includes. It also adds "[sources]" link so that you can get to the complete source of the DAG - including imports etc.

sunank200 replied:
Changed it to `exampleinclude`

amoghrajesh commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,130 @@
+airflow.sdk API Reference
+=========================
+
+This page documents the full public API exposed in Airflow 3.0+ via the Task SDK python module.
```
```suggestion
This page documents the full public API exposed in Airflow 3 and later via the Task SDK python module.
```

sunank200 replied:
Changed it.

amoghrajesh commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,130 @@
+This page documents the full public API exposed in Airflow 3.0+ via the Task SDK python module.
+
+If something is not on this page it is best to assume that it is not part of the public API and use of it is entirely at your own risk
+-- we won't go out of our way break usage of them, but we make no promises either.
```
```suggestion
-- we won't go out of our way and break usage of them, but we make no promises either otherwise.
```

sunank200 replied:
Changed it.

amoghrajesh commented on task-sdk/docs/conf.py:
```diff
@@ -0,0 +1,74 @@
+
+# Prefer pyi over py files if both are found
+autoapi_file_patterns = ["*.pyi", "*.py"]
+# autoapi_generate_api_docs = False
```
Can we remove this?

sunank200 replied:
Removed it.

amoghrajesh commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+
+:any:`DAG` is where to start. :any:`dag`
+
+The Apache Airflow Task SDK(Task SDK) provides python-native interfaces for defining DAGs,
```
```suggestion
The Apache Airflow Task SDK provides python-native interfaces for defining DAGs,
```

sunank200 replied:
Changed it.

amoghrajesh commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+   :language: python
+   :start-after: [START simplest_dag]
+   :end-before: [END simplest_dag]
+   :caption: Simplest DAG with :func:`@dag <airflow.sdk.dag>`  and :func:`@task <airflow.sdk.task>`
```
Is this a repetition from above?

sunank200 replied:
It is just a simple example in the index.rst. This example is also there in the examples.rst

amoghrajesh commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+
+Execution Time Components
+~~~~~~~~~~~~~~~~~~~~~~~~~
+At runtime, tasks run in an isolated subprocess managed by the SDK:
```
```suggestion
At runtime, tasks run in an isolated subprocess managed by the SDK Supervisor
```

sunank200 replied:
Changed it.

amoghrajesh commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+  - **Supervisor** coordinates the worker's lifecycle.
+  - **TaskRunner** actually executes the user's task code.
+  - **Context** objects provide runtime metadata (e.g., connections, variables).
+    (see the **Execution Time** section below for details)
```
If we want to continue mentioning these, we should also mention SDK API client

sunank200 replied:
Added the following: 
```
 **API Client** provides a client for communicating with the Airflow Execution API (e.g., heartbeats, task instance updates, XCom pushes, etc).
```

amoghrajesh commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,102 @@
+
+  - **Supervisor** coordinates the worker's lifecycle.
+  - **TaskRunner** actually executes the user's task code.
+  - **Context** objects provide runtime metadata (e.g., connections, variables).
```
```suggestion
  - **Context** objects provide runtime metadata (e.g., dag run details, connections, variables, etc).
```

sunank200 replied:
changed it.

amoghrajesh commented on task-sdk/docs/spelling_wordlist.txt:
```diff
@@ -0,0 +1,2057 @@
+aarch
```
Do we need this big comprehensive list?

potiuk replied:
Yeah. I think this can be removed - once we switched to the common mechanism `uv run --group docs build-docs` will use the common spelling exclusion file from `airflow/docs` - it is also checked by the quick spell-check we run on pre-commit and it is sorted automatically by another pre-commit.

sunank200 replied:
I’ve removed the entire spelling_wordlist.txt

amoghrajesh commented on task-sdk/tests/test_public_api.py:
```diff
@@ -0,0 +1,74 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
```
I do not know if task-sdk/tests is right for this. Maybe create a `task-sdk/docs`

potiuk replied:
Yeah. I'd say those are good candidates for `pre-commit` - they sound more like static check types than dynamic tests with `pytest`

sunank200 replied:
These public‐api checks are true pytest tests—they import the package at runtime, inspect `__all__`, and assert what’s actually there. 

I think pre-commit hooks in Airflow are intended to be fast, file‐based, static linters or formatters, and they shouldn’t be reaching into the code to execute it.

amoghrajesh replied:
Yeah I think a pytest here makes more sense. But that said, this can be moved to a better location for sure.

sunank200 replied:
Yes I can do it as part of https://github.com/apache/airflow/issues/51519

amoghrajesh reviewed (commented):
Looking nice, some comments

sunank200 committed:
Fix PR comments

amoghrajesh commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,130 @@
+-------------
+.. autoapiclass:: airflow.sdk.DAG
+
+.. autoapifunction:: airflow.sdk.dag
```
This is the dag decorator right? Either move it below into "Decorators" and remove from here or copy it there too.

sunank200 replied:
Moved it to decorator.

amoghrajesh commented on task-sdk/docs/api.rst:
```diff
@@ -0,0 +1,130 @@
+
+Bases
+-----
+.. autoapiclass:: airflow.sdk.BaseOperator
```
We are missing `BaseXCom` here.

sunank200 replied:
Added it

amoghrajesh commented on task-sdk/tests/test_public_api.py:
```diff
@@ -0,0 +1,74 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Maybe in a follow up, move this to a better location, like: `task-sdk/tests/docs`

sunank200 replied:
Yes I can do it as part of https://github.com/apache/airflow/issues/51519

amoghrajesh reviewed (approved):
It handles most of my comments, and the PR need not be perfect in the initial cut. I am OK with this, as it implements the initial skeleton for task SDK docs, so in the interest of setting up the basic layer to build things onto, looks good +1

Please ensure to handle the comments before merge.

sunank200 committed:
Move sections in api.rst

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,103 @@
+Model data as assets and emit them to downstream tasks with the SDK's asset library under
+``airflow.sdk.definitions.asset``. You can use:
+
+- :func:`@asset <airflow.sdk.asset>`, :class:`~airflow.sdk.AssetAlias`, etc. (see the **api reference** section below)
```
This doesn't seem to be linking `@asset` for me.

Not strictly a problem we need to fix in _this_ PR, but I'm surprised it doesn't link, and equally surprised it doesn't cause a warning/error in the build.

sunank200 replied:
I can fix this as part of https://github.com/apache/airflow/issues/51519

sunank200 replied:
Tested it now and it fixes in https://github.com/apache/airflow/pull/51871

ashb commented on task-sdk/docs/index.rst:
```diff
@@ -0,0 +1,103 @@
+  - **TaskRunner** actually executes the user's task code.
+  - **Context** objects provide runtime metadata (e.g., dag run details, connections, variables, etc).
+    (see the **Execution Time** section below for details)
+  - **API Client** provides a client for communicating with the Airflow Execution API (e.g., heartbeats, task instance updates, XCom pushes, etc).
```
```suggestion
```

None of this belongs in user-facing docs. (Maaybe context does, but we don't need a whole section just for that)

sunank200 committed:
Update task-sdk/docs/index.rst

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>
